### ISSUE TITLE
issue with InputLayer + BatchNormalizationLayer using graph api

### ISSUE THREAD

**Issue by ved27** [ARM-software/ComputeLibrary#713]
Hi,

Compute library- 19.05
arm7a
Platform : linux

Output of 'strings libarm_compute.so | grep arm_compute_version':

In the model which I  run using graph api ,
I have InputLayer followed by BatchNormalizationLayer 

I ran the model but I found that data of all the output tensors of nodes is 0 .

After going through the source code , I found that in the InPlaceOperationMutator.cpp the line 60 , 

                **if(new_output_tensor->accessor() == nullptr || current_output_tensor->desc().quant_info == new_output_tensor->desc().quant_info)**

probably causes this issue because , the above condition holds true despite my InputLayer having an accessor . 
The condition should be true only when the accessor is not present AND quant info is same.

Please check if that is a bug on your side , correct me if I'm wrong . 

Thanking you ,

**Comment by ved27** [comment]
Hi ,


Kindly look into the issue 

Thanks,

**Comment by morgolock** [comment]
Hi @ved27 

Could you please share the code to configure the graph?

Is the problem that the final result is just 0s?

**Comment by ved27** [comment]
Hi , 


Not just the final result , output of every intermediate tensor is 0 . Note that I'm not using memory manager .

For some reason I can't share the code .

the network has 

InputLayer -> BatchNormalizationLayer -> ConvolutionLayer -> .................    : output data of all outputs of nodes is zero 

InputLayer -> ConvolutionLayer -> ............ : output data is non zero .


So, 
Can you please let me know if InPlaceOperationMutator.cpp the line 60 
   **if(new_output_tensor->accessor() == nullptr || current_output_tensor->desc().quant_info == new_output_tensor->desc().quant_info)**  is correct ? 



Thanking you ,

**Comment by morgolock** [comment]
Hi @ved27 

How do you check the intermediate results of each layer? What data type are you using FP32 or QASYMM8?

**Comment by ved27** [comment]
Hi @morgolock 

I use fp32 for the network.

First I ran the graph.  After it ran , I iterate over the nodes of the graph to take the specific tensor output based on the name of node and dump the data .

**Comment by morgolock** [comment]
Hi @ved27 

Do you see the same problem in both Neon and OpenCL?

**Comment by ved27** [comment]
Hi ,

Yes .

**Comment by ved27** [comment]
Hi @morgolock 

Any progress on this issue ? 

Regards ,

**Comment by morgolock** [comment]
Hi @ved27 

I'm working on it. There is no easy way to do dump the intermediate results. The graph api always uses the memory manager and this reuses the memory for different tensors in the graph. I'm not completely sure but I think this is why you might see all 0s.

One thing you could is try is modifying an existing accessor to dump the tensors to a numpy file: `arm_compute::utils::save_to_npy(tensor, filename,false);`

https://github.com/ARM-software/ComputeLibrary/blob/master/utils/GraphUtils.cpp

**Comment by ved27** [comment]
Hi @morgolock 

I already set the use_transition_memory_manager to false explicitly .

Please find the following information in one of my above comments , 

```
InputLayer (has an accessor ) -> BatchNormalizationLayer -> ConvolutionLayer -> ................. : output data of all outputs of nodes is zero

InputLayer (has an accessor )-> ConvolutionLayer -> ............ : output data is non zero and CORRECT 
```.

**Comment by morgolock** [comment]
Hi @ved27 

BatchNormalizationLayer is used in a few of our graph examples and we have not noticed any issue with it. 
Please take a look at the way it's configured in this example:
https://github.com/ARM-software/ComputeLibrary/blob/master/examples/graph_inception_v3.cpp#L96
Do you do something similar in your graph?

Have you tried changing the configuration values/data of BatchNormalization for debugging purposes to see if this produces values other than 0? It would be helpful if you could share the configuration values and the data.

**Comment by ved27** [comment]
Hi @morgolock 

Please note that BatchNormalizationLayer is after the InputLayer in my case . Not after ConvolutionLayer like in many examples.

1) InputLayer (has an accessor ) -> BatchNormalizationLayer -> ConvolutionLayer ->  : zero output  
2) InputLayer (has an accessor )-> ConvolutionLayer -> : correct output**


For the above 1) case , please verify the following 
src/graph/mutators/InPlaceOperationMutator.cpp the line 60 , the following condition
if(new_output_tensor->accessor() == nullptr || current_output_tensor->desc().quant_info == new_output_tensor->desc().quant_info ) 

- Please check the code under above if ' condition in src/graph/mutators/InPlaceOperationMutator.cpp

**Comment by IsabellaARM** [comment]
Hi @ved27,
can you give us more information, are you using an image as input? have you tried to change the input values of Batch Normalization? in the meantime, can you try to disable all the mutators and re-run your case 1?
To disable all mutator you can just comment line 72 in src/graph/GraphManager.

```
// Apply all mutating passes
// pm.run_all(graph);
```

**Comment by ved27** [comment]
Hi @IsabellaARM 

I use .npy file as input . 

Disabling mutators results in correct output

**Comment by IsabellaARM** [comment]
Hi @ved27 ,
we have created a patch to fix the problem, can you try it without disabling the mutators?
https://review.mlplatform.org/c/ml/ComputeLibrary/+/1555

It also adds a new accessor that helps to save the intermediate tensors, it's just necessary to add a couple of line in your code, for example:
`graph << ActivationLayer(ActivationLayerInfo(ActivationLayerInfo::ActivationFunction::RELU)).set_name("relu1");
// Add a substream and output layer to save the result of activation on file.npy
SubStream dump_layer(graph);
dump_layer << OutputLayer( get_save_npy_output_accessor("file.npy" );
`

**Comment by ved27** [comment]
Hi @IsabellaARM 

I already changed InPlaceOperationMutator.cpp the line 60 to 
  `if(new_output_tensor->accessor() == nullptr && current_output_tensor->desc().quant_info == new_output_tensor->desc().quant_info)`

and this works .   Since I try running the graph for many loops to check the runtime  , I think the save npy accessor acts as an overhead to the actual runtime . 

Thank you much
